modal/app.py

```diff
@@ -71,58 +70,28 @@ def app_id(self) -> str:
-    ) -> Handle:
-        """Send a server request to create an object in this app, and return its ID."""
-        cached_obj = self._local_uuid_to_object.get(obj.local_uuid)
-        if cached_obj is not None:
-            # We already created this object before, shortcut this method
-            return cached_obj
-
-        resolver = Resolver(self, progress, self._client, self.app_id, existing_object_id)
-
-        # Create object
-        created_obj = await obj._load(resolver)
-
-        resolver.set_finish()  # finishes any message
-
-        if existing_object_id is not None and created_obj.object_id != existing_object_id:
-            # TODO(erikbern): this is a very ugly fix to a problem that's on the server side.
-            # Unlike every other object, images are not assigned random ids, but rather an
-            # id given by the hash of its contents. This means we can't _force_ an image to
-            # have a particular id. The better solution is probably to separate "images"
-            # from "image definitions" or something like that, but that's a big project.
-            if not existing_object_id.startswith("im-"):
-                raise Exception(
-                    f"Tried creating an object using existing id {existing_object_id}"
-                    f" but it has id {created_obj.object_id}"
-                )
-
-        self._local_uuid_to_object[obj.local_uuid] = created_obj
-        return created_obj
-
     async def _create_all_objects(
         self, blueprint: Dict[str, Provider], progress: Tree, new_app_state: int
     ):  # api_pb2.AppState.V
         """Create objects that have been defined but not created on the server."""
+        resolver = Resolver(progress, self._client, self.app_id)
         for tag, provider in blueprint.items():
             existing_object_id = self._tag_to_existing_id.get(tag)
-            self._tag_to_object[tag] = await self._load(provider, progress, existing_object_id)
+            created_obj = await resolver.load(provider, existing_object_id)
+            self._tag_to_object[tag] = created_obj
 
         # Create the app (and send a list of all tagged obs)
         # TODO(erikbern): we should delete objects from a previous version that are no longer needed
         # We just delete them from the app, but the actual objects will stay around
         indexed_object_ids = {tag: obj.object_id for tag, obj in self._tag_to_object.items()}
         unindexed_object_ids = list(
-            set(obj.object_id for obj in self._local_uuid_to_object.values())
+            set(obj.object_id for obj in resolver.objects())
             - set(obj.object_id for obj in self._tag_to_object.values())
         )
         req_set = api_pb2.AppSetObjectsRequest(
             app_id=self._app_id,
             indexed_object_ids=indexed_object_ids,
             unindexed_object_ids=unindexed_object_ids,
-            new_app_state=new_app_state,  # type: ignore
```

Comment:
Ohhh no, I missed this removal in the review :O Was this intentional?

Reply:
Pretty sure it was accidental :(